File: 200_number-of-islands.py
class Solution:
  def numIslands(self, grid: List[List[str]]) -> int:

    def dfs(i: int, j: int) -> None:
      nonlocal grid, n, m
      grid[i][j] = "0"
      for x, y in [[i+1, j], [i-1, j], [i, j+1], [i, j-1]]:
        if 0 <= x < n and 0 <= y < m and grid[x][y] == "1":
          dfs(x, y)

    ans = 0
    n, m = len(grid), len(grid[0])
    for i in range(n):
      for j in range(m):
        if grid[i][j] == "1":
          dfs(i, j)
          ans += 1
    return ans

File: Search/200-number-of-islands.cpp
/******************************************
 * author      : utoppia
 * description : solution for leetcode
 * updated at  : 2024-01-12 20:42:09
 * filename    : 200-number-of-islands.cpp
 * language    : cpp
 * editor      : Emacs 27.1
 *****************************************/
#include <algorithm>
#include <bits/stdc++.h>
#include <ios>
#include <iostream>
#include <set>
#include <utility>
#include <vector>

#define vall(v) v.begin(), v.end()
#define pir pair<int, int>

#define pb push_back
#define eb emplace_back
#define fi first
#define snd second
#define sz size

#define ll long long
#define nMax 200005

#define MOD 998244353
using namespace std;

class Solution {
private:
  void dfs(int x, int y, vector<vector<char>> &grid, int n, int m) {
    if (x < 0 || x >= n || y < 0 || y >= m)
      return;
    if (grid[x][y] == '0')
      return;
    grid[x][y] = '0';
    dfs(x + 1, y, grid, n, m);
    dfs(x - 1, y, grid, n, m);
    dfs(x, y + 1, grid, n, m);
    dfs(x, y - 1, grid, n, m);
  }

public:
  int numIslands(vector<vector<char>> &grid) {
    int n = grid.sz(), m = grid[0].sz();
    int ans = 0;

    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < m; ++j)
        if (grid[i][j] == '1') {
          dfs(i, j, grid, n, m);
          ans++;
        }
    }

    return ans;
  }
};
